// gifbin/gifCard/gifCard.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');
var cx = require('classnames');
var Copy = require('gifbin/copyButton/copyButton.jsx');

var GifCard = React.createClass({
	getDefaultProps: function() {
		return {
			gif : {
				id : '23423545',
				link : "http://i.imgur.com/Mk76H.gif",
				staticLink : "http://i.imgur.com/Mk76Hs.gif",
				favourite : true,
				viewCount : 334,
			}
		};
	},

	getInitialState: function() {
		return {
			hovered: false
		};
	},

	hover : function(){
		this.setState({
			hovered : true
		})
		clearTimeout(this.timeout);
	},

	unhover : function(){
		var self = this;
		this.timeout = setTimeout(function(){
			self.setState({
				hovered : false
			})
		}, 50)
	},


	render : function(){
		var self = this;
		var gif, favourite;
		if(this.state.hovered){
			gif = <img className='gif' src={this.props.gif.link} />
		}

		if(this.props.gif.favourite){
			favourite = <i className='fa fa-heart favourite' />
		}

		return(
			<div className={cx('gifCard', {'hovered' : this.state.hovered})}>
				<img className="static" src={this.props.gif.staticLink} />
				{gif}
				<div className='hitbox' onMouseOver={this.hover} onMouseOut={this.unhover} ref='hitbox'>
					<div className='controls'>
						<Copy link={this.props.gif.link} />
						<a href={'/gif/' + this.props.gif.id} className='editButton'>
							<i className='fa fa-pencil-square-o' />
						</a>
					</div>
				</div>
				<div className='info'>
					{favourite}
					<div className='views'>
						{this.props.gif.viewCount}
					</div>
				</div>

			</div>
		);
	}
});

module.exports = GifCard;
