// gifbin/gifCard/gifCard.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');
var $ = require('jquery');
var cx = require('classnames');
var ClipboardButton = require('gifbin/clipboardButton/clipboardButton.jsx');

var Link = require('gifbin/link.jsx');

var GifCard = React.createClass({
	getDefaultProps: function() {
		return {
			gif : {
				id : '23423545',
				link : "http://i.imgur.com/Mk76H.gif",
				staticLink : "http://i.imgur.com/Mk76Hs.gif",
				favourite : false,
				viewCount : 334,
			}
		};
	},

	getInitialState: function() {
		return {
			hovered: false
		};
	},

	hover : function(){
		clearTimeout(this.timeout);
		if(this.state.hovered) return;
		this.setState({
			hovered : true
		})
	},

	unhover : function(){
		var self = this;
		this.timeout = setTimeout(function(){
			self.setState({
				hovered : false
			})
			if(self.refs){
				$(self.refs.hitbox.getDOMNode()).find('.clipboardButton').removeClass('zeroclipboard-is-hover');
			}
		}, 50)
	},
	componentDidUnmount : function(){
		clearTimeout(this.timeout);
	},


	render : function(){
		var self = this;
		var gifv, favourite;
		if(this.state.hovered){
			//gif = <img className='gif' src={this.props.gif.link} />
			gifv = <video className='gif' autoPlay loop muted>
				<source src={this.props.gif.link.replace('.gif', '.webm')} type="video/webm" />
			</video>
		}


		var heartButton;
		if(this.props.gif.favourite){
			favourite = <i className='fa fa-heart favourite' />
			heartButton = <div className='unfavButton'>
				<i className='fa fa-heart-o' />
			</div>
		}else{
			heartButton = <div className='favButton'>
				<i className='fa fa-heart' />
			</div>
		}

		return(
			<div className={cx('gifCard', {'hovered' : this.state.hovered})}>
				<img className="static" src={this.props.gif.staticLink} />
				{gifv}


				<div className='hitbox' onMouseMove={this.hover} onMouseOut={this.unhover} ref='hitbox'>
					<div className='controls'>
						<ClipboardButton
							className='gifButton'
							icon='fa-link'
							tooltip='gif'
							tooltipDone='copied!'
							link={this.props.gif.link} />

						<ClipboardButton
							className='gifvButton'
							link={this.props.gif.link + 'v'}
							tooltip='gifv'
							tooltipDone='hq copied!'
							icon='fa-rocket' />


						<Link href={'/edit/' + this.props.gif.id} className='editButton'>
							<i className='fa fa-pencil-square-o' />
						</Link>
						{heartButton}
					</div>
				</div>
				<div className='info'>
					{favourite}
					<div className='views'>
						{this.props.gif.viewCount}
					</div>
				</div>

			</div>
		);
	}
});

module.exports = GifCard;
