// gifbin/bucketSelect/bucketSelect.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');
var cx = require('classnames');

var BucketSelect = React.createClass({

	getDefaultProps: function() {
		return {
			buckets : {
				'approve' : {
					name : 'approve',
					img : '/assets/gifbin/bucketSelect/img/noun_49049_cc.png'
				},
				'mind_blown' : {
					name : 'mind blown',
					img : '/assets/gifbin/bucketSelect/img/noun_49050_cc.png'
				},
				'sass' : {
					name : 'sass',
					img : '/assets/gifbin/bucketSelect/img/noun_49054_cc.png'
				},
				'excited' : {
					name : 'excited',
					img : '/assets/gifbin/bucketSelect/img/noun_49045_cc.png'
				},
				'disgust' : {
					name : 'disgust',
					img : '/assets/gifbin/bucketSelect/img/noun_49053_cc.png'
				},
				'frustation' : {
					name : 'frustation',
					img : '/assets/gifbin/bucketSelect/img/noun_49048_cc.png'
				},
				'nope' : {
					name : 'nope',
					img : '/assets/gifbin/bucketSelect/img/noun_49051_cc.png'
				},
				'wat' : {
					name : 'wat',
					img : '/assets/gifbin/bucketSelect/img/noun_49052_cc.png'
				},
				'aw_yis' : {
					name : 'aw yis',
					img : '/assets/gifbin/bucketSelect/img/noun_49047_cc.png'
				},
				'TEST' : {
					name : 'UNKNOWN',
					img : '/assets/gifbin/bucketSelect/img/noun_49046_cc.png'
				}
			},
			selectedBuckets : [],
			onChange : function(){}
		};
	},


	selectBucket : function(bucketId){
		this.props.selectedBuckets.push(bucketId);
		this.props.onChange(this.props.selectedBuckets);
	},

	unselectBucket : function(bucketId){
		this.props.onChange(_.without(this.props.selectedBuckets, bucketId))
	},

	render : function(){
		var self = this;


		var buckets = _.map(this.props.buckets, function(bucket, id){

			var isSelected = _.contains(self.props.selectedBuckets, id);
			var onclick, checkmark;
			if(isSelected){
				onclick = self.unselectBucket.bind(self, id)
				checkmark = <i className='checkmark fa fa-check-circle-o' />
			}else{
				onclick = self.selectBucket.bind(self, id)
			}

			return <div className={cx('bucketItem', {'selected' : isSelected})} onClick={onclick} key={id}>
				<img src={bucket.img} />
				<div className='bucketName'>{bucket.name}</div>
				{checkmark}
			</div>
		})

		return(
			<div className='bucketSelect'>
				{buckets}


			</div>
		);
	}
});

module.exports = BucketSelect;
