// gifbin/searchBar/searchBar.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');
var cx = require('classnames');

var queryVal;

var SearchBar = React.createClass({

	getDefaultProps: function() {
		return {
			onSearch : function(){}
		};
	},

	getInitialState: function() {
		return {
			focused: false,
			showInfoBox : false
		};
	},

	handleChange : function(e){
		this.props.onSearch(e.target.value);
		queryVal = e.target.value;
		this.updateUrl();
	},
	setFocus : function(val){
		this.setState({
			focused : val
		})
	},

	toggleInfoBox : function(){
		this.setState({
			showInfoBox : !this.state.showInfoBox
		})
	},
	hideInfoBox  : function(){
		this.setState({
			showInfoBox : false
		})
	},
	showInfoBox  : function(){
		this.setState({
			showInfoBox : true
		})
	},

	updateUrl : _.throttle(function(){
		var newUrl = window.location.protocol + "//" + document.location.host + document.location.pathname;
		if(queryVal){
			newUrl +="?q=" + queryVal;
		}
		window.history.replaceState(null,null,newUrl);
	}, 500),

	render : function(){
		var self = this;


		var info;
		if(this.state.showInfoBox){
			info = <div className='infoBox' onClick={this.hideInfoBox}>
						Here there, here's some info!

					</div>
		}


		return(
			<div className={cx('searchBar', {'focused' : this.state.focused})}>
				<i className='fa fa-search' />
				<input type='text'
					value={this.props.value}
					onChange={this.handleChange}
					onFocus={this.setFocus.bind(this, true)}
					onBlur={this.setFocus.bind(this, false)} />


				<div className='info'>
					<i className='infoBoxButton fa fa-info' onClick={this.toggleInfoBox} />
					{info}
				</div>


			</div>
		);
	}
});

module.exports = SearchBar;
