// gifbin/copyButton/copyButton.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');

var CopyButton = React.createClass({
	getDefaultProps: function() {
		return {
			link : "http://i.imgur.com/Mk76H.gif",
		};
	},

	getInitialState: function() {
		return {
			tooltipText: 'Copy Link!',
		};
	},

	onLeave : function(){
		this.setState({
			tooltipText : 'Copy Link!'
		});
		var hb = this.refs.hitbox.getDOMNode();
		hb.className = hb.className.replace(/\bzeroclipboard-is-hover\b/,'');
	},

	componentDidMount: function() {
		var self = this;
		var ZeroClipboard = require('zeroclipboard');
		var hitbox = this.refs.hitbox.getDOMNode()
		this.zc = new ZeroClipboard(hitbox);

		this.zc.setText(this.props.link);

		this.zc.on( "aftercopy", function( event ) {
			self.setState({
				tooltipText : 'Copied!'
			});
		} );
	},

	render : function(){
		var self = this;

		return(
			<div className='copyButton' ref='hitbox' data-tooltip={this.state.tooltipText} onMouseLeave={this.onLeave}>
				<i className='fa fa-link' />
			</div>
		);
	}
});

module.exports = CopyButton;
