// gifbin/copyButton/copyButton.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');

var CopyButton = React.createClass({
	getDefaultProps: function() {
		return {
			link : "http://i.imgur.com/Mk76H.gif",
			tooltip : "Copy Link!",
			tooltipDone : "Copied!",
			icon : "fa-link"
		};
	},

	getInitialState: function() {
		return {
			tooltipText: this.props.tooltip,
		};
	},

	onLeave : function(){
		this.setState({
			tooltipText : this.props.tooltip
		});
		var hb = this.refs.hitbox.getDOMNode();
		hb.className = hb.className.replace(/\bzeroclipboard-is-hover\b/,'');
	},

	componentDidMount: function() {
		var self = this;
		var ZeroClipboard = require('zeroclipboard');
		var hitbox = this.refs.hitbox.getDOMNode()
		this.zc = new ZeroClipboard(hitbox);

		this.zc.setText(this.props.link);

		this.zc.on( "aftercopy", function( event ) {
			self.setState({
				tooltipText : self.props.tooltipDone
			});
		} );
	},

	render : function(){
		var self = this;
		return(
			<div className={'copyButton ' + this.props.className} ref='hitbox' data-tooltip-right={this.state.tooltipText} onMouseLeave={this.onLeave}>
				<i className={'fa ' + this.props.icon} />
			</div>
		);
	}
});

module.exports = CopyButton;
