// gifbin/gifstore.js
var Storage = {

	categories : {

	},

	gifs : {
		'23423545' : {
			id : '23423545',
			link : "http://i.imgur.com/Mk76H.gif",
			uploader : "Scott",
			staticLink : "http://i.imgur.com/Mk76Hs.gif",
			favourite : false,
			viewCount : 334,
		},

		'sdfsdff' : {
			id : 'sdfsdff',
			link : "http://i.imgur.com/DxHVTt2.gif",
			staticLink : "http://i.imgur.com/DxHVTt2s.gif",
			uploader : "Scott",
			favourite : true,
			viewCount : 12
		},
		'23423545f' : {
			id : '23423545f',
			link : "http://i.imgur.com/Mk76H.gif",
			uploader : "Scott",
			staticLink : "http://i.imgur.com/Mk76Hs.gif",
			favourite : false,
			viewCount : 334,
		},

		'sdfsdfg' : {
			id : 'sdfsdfg',
			link : "http://i.imgur.com/DxHVTt2.gif",
			staticLink : "http://i.imgur.com/DxHVTt2s.gif",
			uploader : "Scott",
			favourite : true,
			viewCount : 12
		},
		'23423545g' : {
			id : '23423545g',
			link : "http://i.imgur.com/Mk76H.gif",
			uploader : "Scott",
			staticLink : "http://i.imgur.com/Mk76Hs.gif",
			favourite : false,
			viewCount : 334,
		},

		'sdfsdfh' : {
			id : 'sdfsdfh',
			link : "http://i.imgur.com/DxHVTt2.gif",
			staticLink : "http://i.imgur.com/DxHVTt2s.gif",
			uploader : "Scott",
			favourite : true,
			viewCount : 12
		},
		'23423545h' : {
			id : '23423545h',
			link : "http://i.imgur.com/Mk76H.gif",
			uploader : "Scott",
			staticLink : "http://i.imgur.com/Mk76Hs.gif",
			favourite : false,
			viewCount : 334,
		},

		'sdfsdfj' : {
			id : 'sdfsdfj',
			link : "http://i.imgur.com/DxHVTt2.gif",
			staticLink : "http://i.imgur.com/DxHVTt2s.gif",
			uploader : "Scott",
			favourite : true,
			viewCount : 12
		}
	},

	user : {

	},

	avatars : {
		'Scott' : '23423545'
	}

}










module.exports = {

	getGifs : function(){
		return Storage.gifs;
	},

};
