// gifbin/utils.js
var _ = require('underscore');


module.exports = {

	getQuery : function(){
		if(typeof window === 'undefined') return;

		if(document.location.search){
			var path = document.location.search;
			var query = _.object(_.map(path.substring(path.indexOf('?')+1).split('&'), function(part){
				return part.split('=');
			}));

			return decodeURIComponent(query.q);
		}
	},

	getImageFromUrl : function(){
		if(typeof window === 'undefined') return;

		if(document.location.search){
			var path = document.location.search;
			var query = _.object(_.map(path.substring(path.indexOf('?')+1).split('&'), function(part){
				return part.split('=');
			}));

			return decodeURIComponent(query.i);
		}
	},

	createSearchObject : function(query){
		var result = {
			buckets : [],
			users : [],
			tags : [],
			query : query
		};

		if(!query) return result;

		var searchTerms = _.chain(query.toLowerCase().split(' '))
			.map(function(term){
				return term.split(',')
			})
			.flatten()
			.filter(function(term){
				return term !== "";
			})
			.value();

		_.each(searchTerms, function(term){
			if(term.indexOf('by:') === 0){
				result.users.push(term.substring(3));
			}else if(term.indexOf('in:') === 0){
				result.buckets.push(term.substring(3));
			}else{
				result.tags.push(term)
			}
		});

		return result;
	},



}
