// gifbin/gifContainer/gifContainer.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');

var GifStore = require('gifbin/gif.store.js');
var GifCard = require('gifbin/gifCard/gifCard.jsx');

var GifContainer = React.createClass({

/*
	mixins : [GifStore.mixin()],

	onStoreChange  : function(){
		this.setState({
			featured: GifStore.getFeaturedState(),
			all: GifStore.getAppsState(),
		});
	},
*/

	getDefaultProps: function() {
		return {
			searchObj : {}
		};
	},

	makeUserResults : function(gifs){
		var searchObj = this.props.searchObj;
		var userText = searchObj.users.join(', ');

		if(searchObj.buckets.length){
			userText += ' in ' + searchObj.buckets.join(', ')
		}

		return <div className='gifCollection'>
			<GifBox title={"Fav by " + userText} gifs={gifs.favouritedBy} />
			<GifBox title={"Uploaded by " + userText} gifs={gifs.uploadedBy} />
		</div>

	},


	render : function(){
		var self = this;
		var searchObj = this.props.searchObj;

		var gifs = GifStore.search(this.props.searchObj);


		if(searchObj.users.length){
			return this.makeUserResults(gifs);
		}

		var bucketText = "";
		if(searchObj.buckets.length){
			bucketText = searchObj.buckets.join(', ')
		}

		return <GifBox title={bucketText} gifs={gifs.all} />
	}
});


var GifBox = React.createClass({

	render : function(){
		var self = this;

		var header;
		if(this.props.title){
			header = <h1>{this.props.title}</h1>
		}

		var gifs = _.map(this.props.gifs, function(gif){
			return <GifCard gif={gif} key={gif.id} />
		});

		return(
			<div className='gifContainer'>
				{header}
				{gifs}
			</div>
		);
	}
});


module.exports = GifContainer;
