// gifbin/clipboardButton/clipboardButton.jsx
/** @jsx React.DOM */
var React = require('react');
var _ = require('underscore');
var $ = require('jquery');

var ClipboardButton = React.createClass({
	getDefaultProps: function() {
		return {
			link : "http://i.imgur.com/Mk76H.gif",
			tooltip : "Copy Link!",
			tooltipDone : "Copied!",
			icon : "fa-link",
			tooltipDirection : 'right'
		};
	},

	getInitialState: function() {
		return {
			tooltipText: this.props.tooltip,
		};
	},

	onLeave : function(){
		this.setState({
			tooltipText : this.props.tooltip
		});
		$(this.refs.hitbox.getDOMNode()).removeClass('zeroclipboard-is-hover');
		this.zc.destroy();
	},

	onEnter : function(){
		this.setupZC();
	},

	setupZC : function(){
		var self = this;
		var ZeroClipboard = require('zeroclipboard');
		var hitbox = this.refs.hitbox.getDOMNode()
		this.zc = new ZeroClipboard(hitbox);
		ZeroClipboard.config( { swfPath: "/assets/gifbin/clipboardButton/ZeroClipboard.swf" } )

		this.zc.setText(this.props.link);

		this.zc.on( "aftercopy", function( event ) {
			self.setState({
				tooltipText : self.props.tooltipDone
			});
		} );
	},

	render : function(){
		var self = this;

		var extraProps = {};

		if(this.props.tooltip && this.props.tooltipDirection){
			extraProps['data-tooltip-' + this.props.tooltipDirection] = this.state.tooltipText;
		}

		return(
			<div {...extraProps} className={'clipboardButton ' + this.props.className}
					ref='hitbox'  onMouseLeave={this.onLeave}
					onMouseEnter={this.onEnter}>
				<i className={'fa ' + this.props.icon} />
			</div>
		);
	}
});

module.exports = ClipboardButton;
